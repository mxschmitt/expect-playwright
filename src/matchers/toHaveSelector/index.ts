import { SyncExpectationResult } from "expect/build/types"
import type { Page } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../../global"

const toHaveSelector: jest.CustomMatcher = async function (
  page: Page,
  selector: string,
  options: PageWaitForSelectorOptions = {}
): Promise<SyncExpectationResult> {
  const pass = await page
    .waitForSelector(selector, {
      state: this.isNot ? "hidden" : "visible",
      ...options,
    })
    .then(() => !this.isNot)
    .catch(() => this.isNot)

  return {
    pass: pass,
    message: () => {
      const not = this.isNot ? " not" : ""
      const hint = this.utils.matcherHint(
        "toHaveSelector",
        undefined,
        undefined,
        { isNot: this.isNot, promise: this.promise }
      )

      return (
        hint +
        "\n\n" +
        `Expected: page to${not} have selector ${this.utils.printExpected(
          selector
        )}`
      )
    },
  }
}

export default toHaveSelector
