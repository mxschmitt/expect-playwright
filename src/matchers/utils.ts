import type { Page, ElementHandle, Frame } from "playwright-core"
import { PageWaitForSelectorOptions } from "../../global"

export type ExpectInputType = Page | Frame | ElementHandle

const isElementHandle = (value: ExpectInputType): value is ElementHandle => {
  return value.constructor.name === "ElementHandle"
}

export const getFrame = async (value: ExpectInputType) => {
  return isElementHandle(value) ? value.contentFrame() : value
}

const isObject = (value: unknown) =>
  typeof value === "object" && !(value instanceof RegExp)

export type InputArguments = [
  Page | ElementHandle,
  string?,
  (string | PageWaitForSelectorOptions)?,
  PageWaitForSelectorOptions?
]

export const getElementHandle = async (
  args: InputArguments,
  valueArgCount = 1
) => {
  // Pluck the options off the end first
  const options =
    args.length > 1 && isObject(args[args.length - 1])
      ? (args.pop() as PageWaitForSelectorOptions)
      : {}

  // Next, pluck the number of args required by the matcher (defaults to 1)
  const expectedValue = args.splice(-valueArgCount, valueArgCount) as string[]

  // Finally, we can find the element handle
  let elementHandle = (await getFrame(args[0])) ?? args[0]

  // If the user provided a page or iframe, we need to locate the provided
  // selector or the `body` element if none was provided.
  if (!isElementHandle(elementHandle)) {
    const selector = args[1] ?? "body"

    try {
      elementHandle = (await elementHandle.waitForSelector(selector, options))!
    } catch (err) {
      throw new Error(`Timeout exceed for element ${quote(selector)}`)
    }
  }

  return [elementHandle, expectedValue] as const
}

export const quote = (val: string | null) => (val === null ? "" : `'${val}'`)

export const getMessage = (
  { isNot, promise, utils, expand }: jest.MatcherContext,
  matcher: string,
  expected: unknown,
  received: unknown,
  expectedHint: string | undefined = undefined
) => {
  const message = isNot
    ? `Expected: not ${utils.printExpected(expected)}`
    : utils.printDiffOrStringify(
        expected,
        received,
        "Expected",
        "Received",
        expand
      )

  return (
    utils.matcherHint(matcher, undefined, expectedHint, { isNot, promise }) +
    "\n\n" +
    message
  )
}

export const compareText = (
  expectedValue: string | RegExp,
  actualValue: string | null
) => {
  return typeof expectedValue === "string"
    ? expectedValue === actualValue
    : expectedValue.test(actualValue ?? "")
}
