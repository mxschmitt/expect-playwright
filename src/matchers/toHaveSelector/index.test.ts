import { assertSnapshot } from "../tests/utils"
import toHaveSelector from "."

expect.extend({ toHaveSelector })

describe("toHaveSelector", () => {
  afterEach(async () => {
    await page.setContent("")
  })
  it("positive", async () => {
    await page.setContent(`<div id="foobar">Bar</div>`)
    await expect(page).toHaveSelector("#foobar")
  })
  it("negative", async () => {
    await assertSnapshot(() =>
      expect(page).toHaveSelector("#foobar", { timeout: 1000 })
    )
  })

  describe("with 'not' usage", () => {
    it("positive", async () => {
      await expect(page).not.toHaveSelector("#foobar", { timeout: 1000 })
    })

    it("negative", async () => {
      await page.setContent(`<div id="foobar">Bar</div>`)
      await assertSnapshot(() => expect(page).not.toHaveSelector("#foobar"))
    })
  })

  describe("timeout", () => {
    it("positive: should be able to use a custom timeout", async () => {
      setTimeout(async () => {
        await page.setContent(`<div id="foobar">Bar</div>`)
      }, 500)
      await expect(page).toHaveSelector("#foobar", { timeout: 1000 })
    })
    it("should throw an error after the timeout exceeds", async () => {
      const start = new Date().getTime()
      await assertSnapshot(() =>
        expect(page).toHaveSelector("#foobar", { timeout: 1000 })
      )
      const duration = new Date().getTime() - start
      expect(duration).toBeLessThan(1500)
    })
  })
})
