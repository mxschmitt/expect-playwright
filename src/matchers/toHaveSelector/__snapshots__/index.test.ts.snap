// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveSelector negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoHaveSelector[2m([22m[32mexpected[39m[2m)[22m

Expected: page to have selector [32m\\"#foobar\\"[39m"
`;

exports[`toHaveSelector timeout should throw an error after the timeout exceeds 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mtoHaveSelector[2m([22m[32mexpected[39m[2m)[22m

Expected: page to have selector [32m\\"#foobar\\"[39m"
`;

exports[`toHaveSelector with 'not' usage negative 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).[22mnot[2m.[22mtoHaveSelector[2m([22m[32mexpected[39m[2m)[22m

Expected: page to not have selector [32m\\"#foobar\\"[39m"
`;
